File: 189-rotate-array/189-rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        if( k == 0 || k == nums.size() || nums.size() < 2 )
            return;
        int n = nums.size();
        int i = 0;
        int count = 0;
        int hold,currK,temp;
        while(count != n)
        {
            hold = nums[i];
            currK = i;
            do
            {
                temp = (currK+k)%n;
                swap(hold,nums[temp]);
                ++count;
                currK = temp;
            }while(currK != i);
            ++i;
        }
    }
};

File: 189-rotate-array/189-rotate-array.java
class Solution {
    public void rotate(int[] nums, int k) {
        if (nums.length == 1 || nums.length == 0) {
            return;
        }
        k = k % nums.length;                                                             
    //length of first part                                                          
        int a = nums.length - k; 
        reverse(nums, 0, a-1);
        reverse(nums, a, nums.length-1);
        reverse(nums, 0, nums.length-1);
    }
    
    void reverse(int[] nums, int s, int e) {
        while(s<e) {
            int temp=nums[s];
            nums[s]=nums[e];
            nums[e] = temp;
            s++;e--;
        }
    }
}
